### PR TITLE
[GLUTEN-6148] Simplify JniLibLoader loading mechanism for native libraries

### PR DESCRIPTION
**Pull request by ArnavBalyan**
## What changes were proposed in this pull request?
 - Currently Gluten uses an overdesigned library loading mechanism consisting of rollbacks and transactional loading of native libraries.
 - However during failures, an exception is thrown and app simply crashes without need for "rollbacks".
 - Simplified this design, now we simply load libraries sequentially and maintain the existing behaviour of throwing error when there is an issue with loading the libraries. (This intentionally does not change signature of existing used methods of `JniLibLoader` to maintain simplicity in the change)

Fixes: #6148 

## How was this patch tested?
 - Local Spark shell and ran some SparkSQL queries with gluten.
 - Unit tests should cover the rest, needed some feedback on how we can test it better

### CODE REVIEW

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java [hunk 1]
```diff
@@ -175,160 +150,80 @@ public static synchronized void unloadFromPath(String libPath) {
...
-    public JniLoadTransaction mapAndLoad(String unmappedLibName, boolean requireUnload) {
-      try {
-        final String mappedLibName = System.mapLibraryName(unmappedLibName);
-        load(mappedLibName, requireUnload);
-        return this;
-      } catch (Exception e) {
-        abort();
-        throw new GlutenException(e);
+  public void loadAndCreateLink(String libName, String linkName, boolean requireUnload) {
+    sync.lock();
```
Comment (zhztheplayer) [inline gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:184]: Do we still need this lock as class member? Perhaps we can minimize the lock scope by only making operations on `loadedLibraries` atomic. Say `synchronized (loadedLibraries) {...}` may be enough.

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java [hunk 2]
```diff
@@ -72,7 +62,7 @@ public class JniLibLoader {
   private final Set<String> loadedLibraries = new HashSet<>();
   private final Lock sync = new ReentrantLock();
 
-  JniLibLoader(String workDir) {
+  public JniLibLoader(String workDir) {
```
Comment (zhztheplayer) [inline gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java:65]: Can we keep the ctor package-private? If no other places are calling it.

**Review by zhztheplayer — APPROVED** [review]
Just a little minors. Thanks.

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java (hunk 1 above)
Comment (ArnavBalyan) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java]: That's a good catch, updated thanks

File: gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java (hunk 2 above)
Comment (ArnavBalyan) [reply gluten-core/src/main/java/org/apache/gluten/vectorized/JniLibLoader.java]: Done thanks!

**Review by zhztheplayer — APPROVED** [review]
I see there are still some unnecessary synchronizations but should be benign overall. So let's merge this first. Thank you for the effort.

**Comment by ArnavBalyan** [comment]
> I see there are still some unnecessary synchronizations but should be benign overall. So let's merge this first. Thank you for the effort.

Ah missed the inner ones, I'll remove in a follow up change

**Pull request merged**